Serialization issue when using windowed aggregation in Pipeline API

```
com.hazelcast.jet.JetException: Exception in ProcessorTasklet{vertex=listSink(sink), processor=com.hazelcast.jet.impl.pipeline.FunctionAdapter$AdaptingProcessor@60cfb273}: com.hazelcast.nio.serialization.HazelcastSerializationException: Failed to serialize 'com.hazelcast.jet.datamodel.TimestampedItem'
    at com.hazelcast.jet.impl.execution.TaskletExecutionService$BlockingWorker.run(TaskletExecutionService.java:196)
    at java.util.concurrent.Executors$RunnableAdapter.call$$$capture(Executors.java:511)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java)
    at java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:266)
    at java.util.concurrent.FutureTask.run(FutureTask.java)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
Caused by: com.hazelcast.nio.serialization.HazelcastSerializationException: Failed to serialize 'com.hazelcast.jet.datamodel.TimestampedItem'
    at com.hazelcast.internal.serialization.impl.SerializationUtil.handleSerializeException(SerializationUtil.java:75)
    at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:161)
    at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:137)
    at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toData(AbstractSerializationService.java:122)
    at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toData(AbstractSerializationService.java:110)
    at com.hazelcast.spi.impl.NodeEngineImpl.toData(NodeEngineImpl.java:307)
    at com.hazelcast.collection.impl.collection.AbstractCollectionProxyImpl.addAll(AbstractCollectionProxyImpl.java:166)
    at com.hazelcast.jet.impl.connector.HazelcastWriters.lambda$null$3c5c3833$1(HazelcastWriters.java:205)
    at com.hazelcast.jet.impl.connector.WriteBufferedP.process(WriteBufferedP.java:77)
    at com.hazelcast.jet.impl.processor.ProcessorWrapper.process(ProcessorWrapper.java:49)
    at com.hazelcast.jet.impl.pipeline.FunctionAdapter$AdaptingProcessor.process(FunctionAdapter.java:126)
    at com.hazelcast.jet.impl.execution.ProcessorTasklet.stateMachineStep(ProcessorTasklet.java:185)
    at com.hazelcast.jet.impl.execution.ProcessorTasklet.stateMachineStep(ProcessorTasklet.java:156)
    at com.hazelcast.jet.impl.execution.ProcessorTasklet.call(ProcessorTasklet.java:143)
    at com.hazelcast.jet.impl.execution.ProcessorTasklet.call(ProcessorTasklet.java:136)
    at com.hazelcast.jet.impl.execution.TaskletExecutionService$BlockingWorker.run(TaskletExecutionService.java:185)
    ... 7 more
Caused by: java.io.NotSerializableException: com.hazelcast.jet.datamodel.TwoBags
    at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1184)
    at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
    at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
    at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
    at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
    at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:348)
    at com.hazelcast.internal.serialization.impl.JavaDefaultSerializers$JavaSerializer.write(JavaDefaultSerializers.java:111)
    at com.hazelcast.internal.serialization.impl.JavaDefaultSerializers$JavaSerializer.write(JavaDefaultSerializers.java:104)
    at com.hazelcast.internal.serialization.impl.StreamSerializerAdapter.write(StreamSerializerAdapter.java:43)
    at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:158)
    ... 21 more
```

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/datamodel/DataModelSerializerHooks.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/datamodel/DataModelSerializerHooks.java
@@ -39,6 +39,48 @@ import static com.hazelcast.jet.datamodel.TwoBags.twoBags;
  * com.hazelcast.jet.datamodel} package. This is not a public-facing API.
  */
 class DataModelSerializerHooks {
+
+    public static final class TimestampedItemHook implements SerializerHook<TimestampedItem> {
+
+        @Override
+        public Class<TimestampedItem> getSerializationType() {
+            return TimestampedItem.class;
+        }
+
+        @Override
+        public Serializer createSerializer() {
+            return new StreamSerializer<TimestampedItem>() {
+                @Override
+                public int getTypeId() {
+                    return SerializerHookConstants.TIMESTAMPED_ITEM;
+                }
+
+                @Override
+                public void destroy() {
+
+                }
+
+                @Override
+                public void write(ObjectDataOutput out, TimestampedItem timestampedItem) throws IOException {
+                    out.writeLong(timestampedItem.timestamp());
+                    out.writeObject(timestampedItem.item());
+
+                }
+
+                @Override
+                public TimestampedItem read(ObjectDataInput in) throws IOException {
+                    long timestamp = in.readLong();
+                    Object item = in.readObject();
+                    return new TimestampedItem<>(timestamp, item);
+                }
+            };
+        }
+
+        @Override public boolean isOverwritable() {
+            return false;
+        }
+    }
+
     public static final class TimestampedEntryHook implements SerializerHook<TimestampedEntry> {
 
         @Override

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/serialization/SerializerHookConstants.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/serialization/SerializerHookConstants.java
@@ -52,6 +52,7 @@ public final class SerializerHookConstants {
     public static final int HASH_MAP = -322;
     public static final int HASH_SET = -323;
     public static final int JET_EVENT_IMPL = -324;
+    public static final int TIMESTAMPED_ITEM = -325;
 
     // reserved for hadoop module: -380 to -390
 

--- a/hazelcast-jet-core/src/main/resources/META-INF/services/com.hazelcast.SerializerHook
+++ b/hazelcast-jet-core/src/main/resources/META-INF/services/com.hazelcast.SerializerHook
@@ -6,6 +6,7 @@ com.hazelcast.jet.accumulator.AccumulatorSerializerHooks$LongLongAccHook
 com.hazelcast.jet.accumulator.AccumulatorSerializerHooks$LongDoubleAccHook
 com.hazelcast.jet.core.CoreSerializerHooks$WatermarkHook
 com.hazelcast.jet.datamodel.DataModelSerializerHooks$TimestampedEntryHook
+com.hazelcast.jet.datamodel.DataModelSerializerHooks$TimestampedItemHook
 com.hazelcast.jet.datamodel.DataModelSerializerHooks$WindowResultHook
 com.hazelcast.jet.datamodel.DataModelSerializerHooks$Tuple2Hook
 com.hazelcast.jet.datamodel.DataModelSerializerHooks$Tuple3Hook

--- a/hazelcast-jet-core/src/test/java/com/hazelcast/jet/impl/pipeline/transform/AggregateTransform_IntegrationTest.java
+++ b/hazelcast-jet-core/src/test/java/com/hazelcast/jet/impl/pipeline/transform/AggregateTransform_IntegrationTest.java
@@ -16,59 +16,181 @@
 
 package com.hazelcast.jet.impl.pipeline.transform;
 
-import com.hazelcast.config.EventJournalConfig;
 import com.hazelcast.core.IMap;
+import com.hazelcast.jet.IListJet;
 import com.hazelcast.jet.JetInstance;
-import com.hazelcast.jet.config.JetConfig;
+import com.hazelcast.jet.aggregate.AggregateOperation;
 import com.hazelcast.jet.core.JetTestSupport;
+import com.hazelcast.jet.datamodel.Tag;
+import com.hazelcast.jet.datamodel.ThreeBags;
+import com.hazelcast.jet.datamodel.TwoBags;
+import com.hazelcast.jet.pipeline.AggregateBuilder;
+import com.hazelcast.jet.pipeline.BatchStage;
 import com.hazelcast.jet.pipeline.Pipeline;
 import com.hazelcast.jet.pipeline.Sinks;
 import com.hazelcast.jet.pipeline.Sources;
-import com.hazelcast.spi.properties.GroupProperty;
 import com.hazelcast.test.HazelcastParallelClassRunner;
 import com.hazelcast.test.annotation.ParallelTest;
-import org.junit.Before;
+import java.util.Collections;
+import java.util.List;
 import org.junit.Test;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
 
 import static com.hazelcast.jet.Util.entry;
 import static com.hazelcast.jet.aggregate.AggregateOperations.toSet;
+import static com.hazelcast.jet.aggregate.AggregateOperations.toThreeBags;
+import static com.hazelcast.jet.aggregate.AggregateOperations.toTwoBags;
 import static com.hazelcast.jet.core.TestUtil.set;
 import static com.hazelcast.jet.core.test.TestSupport.listToString;
+import static com.hazelcast.jet.datamodel.ThreeBags.threeBags;
 import static java.util.Collections.singletonList;
 import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
 
 @Category(ParallelTest.class)
 @RunWith(HazelcastParallelClassRunner.class)
 public class AggregateTransform_IntegrationTest extends JetTestSupport {
 
-    private JetInstance instance;
-
-    @Before
-    public void before() {
-        JetConfig config = new JetConfig();
-        config.getHazelcastConfig().addEventJournalConfig(
-                new EventJournalConfig().setMapName("source").setEnabled(true));
-        config.getHazelcastConfig().setProperty(GroupProperty.PARTITION_COUNT.getName(), "6");
-        instance = createJetMember(config);
-    }
-
     @Test
-    public void test() {
+    public void test_aggregate() {
+        // Given
+        JetInstance instance = createJetMember();
         IMap<Long, String> map = instance.getMap("source");
         map.put(0L, "foo");
         map.put(1L, "bar");
 
+        // When
         Pipeline p = Pipeline.create();
         p.drawFrom(Sources.<Long, String>map("source"))
          .aggregate(toSet())
          .drainTo(Sinks.list("sink"));
-
         instance.newJob(p).join();
 
+        //Then
         assertEquals(
                 listToString(singletonList(set(entry(0L, "foo"), entry(1L, "bar")))),
                 listToString(instance.getHazelcastInstance().getList("sink")));
+    }
+
+    @Test
+    public void test_aggregate2() {
+        // Given
+        JetInstance instance = createJetMember();
+        IListJet<Integer> list = instance.getList("list");
+        list.add(1);
+        list.add(2);
+        list.add(3);
+        IListJet<String> list1 = instance.getList("list1");
+        list1.add("a");
+        list1.add("b");
+        list1.add("c");
+
+        // When
+        Pipeline p = Pipeline.create();
+        BatchStage<Integer> stage1 = p.drawFrom(Sources.list("list1"));
+        p.drawFrom(Sources.list("list"))
+         .aggregate2(stage1, toTwoBags())
+         .drainTo(Sinks.list("sink"));
+        instance.newJob(p).join();
+
+        //Then
+        TwoBags twoBags = (TwoBags) instance.getHazelcastInstance().getList("sink").iterator().next();
+        assertNotNull(twoBags);
+        sort(twoBags);
+        assertEquals(TwoBags.twoBags(list, list1), twoBags);
+    }
+
+    @Test
+    public void test_aggregate3() {
+        // Given
+        JetInstance instance = createJetMember();
+        IListJet<Integer> list = instance.getList("list");
+        list.add(1);
+        list.add(2);
+        list.add(3);
+        IListJet<String> list1 = instance.getList("list1");
+        list1.add("a");
+        list1.add("b");
+        list1.add("c");
+        IListJet<Double> list2 = instance.getList("list2");
+        list2.add(6.0d);
+        list2.add(7.0d);
+        list2.add(8.0d);
+
+        // When
+        Pipeline p = Pipeline.create();
+        BatchStage<String> stage1 = p.drawFrom(Sources.list("list1"));
+        BatchStage<Double> stage2 = p.drawFrom(Sources.list("list2"));
+        p.drawFrom(Sources.list("list"))
+         .aggregate3(stage1, stage2, toThreeBags())
+         .drainTo(Sinks.list("sink"));
+        instance.newJob(p).join();
+
+        //Then
+        ThreeBags threeBags = (ThreeBags) instance.getHazelcastInstance().getList("sink").iterator().next();
+        assertNotNull(threeBags);
+        sort(threeBags);
+        assertEquals(threeBags(list, list1, list2), threeBags);
+    }
+
+
+    @Test
+    public void test_aggregate3_with_aggBuilder() {
+        // Given
+        JetInstance instance = createJetMember();
+        IListJet<Integer> list = instance.getList("list");
+        list.add(1);
+        list.add(2);
+        list.add(3);
+        IListJet<String> list1 = instance.getList("list1");
+        list1.add("a");
+        list1.add("b");
+        list1.add("c");
+        IListJet<Double> list2 = instance.getList("list2");
+        list2.add(6.0d);
+        list2.add(7.0d);
+        list2.add(8.0d);
+
+        // When
+        Pipeline p = Pipeline.create();
+        BatchStage<Integer> stage0 = p.drawFrom(Sources.list("list"));
+        BatchStage<String> stage1 = p.drawFrom(Sources.list("list1"));
+        BatchStage<Double> stage2 = p.drawFrom(Sources.list("list2"));
+
+
+        AggregateBuilder<Integer> builder = stage0.aggregateBuilder();
+        Tag<Integer> tag0 = builder.tag0();
+        Tag<String> tag1 = builder.add(stage1);
+        Tag<Double> tag2 = builder.add(stage2);
+
+        BatchStage<ThreeBags> resultStage = builder.build(AggregateOperation
+                .withCreate(ThreeBags::threeBags)
+                .andAccumulate(tag0, (acc, item0) -> acc.bag0().add(item0))
+                .andAccumulate(tag1, (acc, item1) -> acc.bag1().add(item1))
+                .andAccumulate(tag2, (acc, item2) -> acc.bag2().add(item2))
+                .andCombine(ThreeBags::combineWith)
+                .andDeduct(ThreeBags::deduct)
+                .andFinish(ThreeBags::finish));
+
+        resultStage.drainTo(Sinks.list("sink"));
+        instance.newJob(p).join();
+
+        //Then
+        ThreeBags threeBags = (ThreeBags) instance.getHazelcastInstance().getList("sink").iterator().next();
+        assertNotNull(threeBags);
+        sort(threeBags);
+        assertEquals(threeBags(list, list1, list2), threeBags);
+    }
+
+    public void sort(ThreeBags threeBags) {
+        Collections.sort((List) threeBags.bag0());
+        Collections.sort((List) threeBags.bag1());
+        Collections.sort((List) threeBags.bag2());
+    }
+
+    public void sort(TwoBags twoBags) {
+        Collections.sort((List) twoBags.bag0());
+        Collections.sort((List) twoBags.bag1());
     }
 }

--- a/hazelcast-jet-core/src/test/java/com/hazelcast/jet/impl/pipeline/transform/WindowAggregateTransform_IntegrationTest.java
+++ b/hazelcast-jet-core/src/test/java/com/hazelcast/jet/impl/pipeline/transform/WindowAggregateTransform_IntegrationTest.java
@@ -19,29 +19,33 @@ package com.hazelcast.jet.impl.pipeline.transform;
 import com.hazelcast.config.EventJournalConfig;
 import com.hazelcast.core.IMap;
 import com.hazelcast.jet.JetInstance;
-import com.hazelcast.jet.pipeline.JournalInitialPosition;
 import com.hazelcast.jet.config.JetConfig;
 import com.hazelcast.jet.core.JetTestSupport;
+import com.hazelcast.jet.datamodel.ThreeBags;
 import com.hazelcast.jet.datamodel.TimestampedItem;
+import com.hazelcast.jet.datamodel.TwoBags;
 import com.hazelcast.jet.datamodel.WindowResult;
 import com.hazelcast.jet.pipeline.Pipeline;
 import com.hazelcast.jet.pipeline.Sinks;
 import com.hazelcast.jet.pipeline.Sources;
+import com.hazelcast.jet.pipeline.StreamStage;
 import com.hazelcast.jet.pipeline.WindowDefinition;
 import com.hazelcast.spi.properties.GroupProperty;
 import com.hazelcast.test.HazelcastParallelClassRunner;
 import com.hazelcast.test.annotation.ParallelTest;
+import java.util.Map.Entry;
 import org.junit.Before;
 import org.junit.Test;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
 
-import java.util.Map.Entry;
-
 import static com.hazelcast.jet.Util.entry;
 import static com.hazelcast.jet.aggregate.AggregateOperations.toSet;
+import static com.hazelcast.jet.aggregate.AggregateOperations.toThreeBags;
+import static com.hazelcast.jet.aggregate.AggregateOperations.toTwoBags;
 import static com.hazelcast.jet.core.TestUtil.set;
 import static com.hazelcast.jet.core.test.TestSupport.listToString;
+import static com.hazelcast.jet.pipeline.JournalInitialPosition.START_FROM_OLDEST;
 import static java.util.Arrays.asList;
 import static org.junit.Assert.assertEquals;
 
@@ -55,7 +59,7 @@ public class WindowAggregateTransform_IntegrationTest extends JetTestSupport {
     public void before() {
         JetConfig config = new JetConfig();
         config.getHazelcastConfig().addEventJournalConfig(
-                new EventJournalConfig().setMapName("source").setEnabled(true));
+                new EventJournalConfig().setMapName("source*").setEnabled(true));
         config.getHazelcastConfig().setProperty(GroupProperty.PARTITION_COUNT.getName(), "1");
         instance = createJetMember(config);
     }
@@ -70,7 +74,7 @@ public class WindowAggregateTransform_IntegrationTest extends JetTestSupport {
         map.put(10L, "flush-item");
 
         Pipeline p = Pipeline.create();
-        p.drawFrom(Sources.<Long, String>mapJournal("source", JournalInitialPosition.START_FROM_OLDEST))
+        p.drawFrom(Sources.<Long, String>mapJournal("source", START_FROM_OLDEST))
          .addTimestamps(Entry::getKey, 0)
          .window(WindowDefinition.tumbling(2))
          .aggregate(toSet())
@@ -95,7 +99,7 @@ public class WindowAggregateTransform_IntegrationTest extends JetTestSupport {
         map.put(10L, "flush-item");
 
         Pipeline p = Pipeline.create();
-        p.drawFrom(Sources.<Long, String>mapJournal("source", JournalInitialPosition.START_FROM_OLDEST))
+        p.drawFrom(Sources.<Long, String>mapJournal("source", START_FROM_OLDEST))
          .addTimestamps(Entry::getKey, 0)
          .window(WindowDefinition.session(2))
          .aggregate(toSet(), (winStart, winEnd, result) -> new WindowResult<>(winStart, winEnd, "", result))
@@ -110,5 +114,103 @@ public class WindowAggregateTransform_IntegrationTest extends JetTestSupport {
                             new WindowResult<>(3, 6, "", set(entry(3L, "bar"), entry(4L, "baz"))))),
                     listToString(instance.getHazelcastInstance().getList("sink")));
         }, 5);
+    }
+
+    @Test
+    public void test_aggregate2() {
+        IMap<Long, String> map = instance.getMap("source");
+        // key is timestamp
+        map.put(0L, "foo");
+        map.put(2L, "baz");
+        map.put(10L, "flush-item");
+
+        IMap<Long, String> map2 = instance.getMap("source1");
+        // key is timestamp
+        map2.put(0L, "faa");
+        map2.put(2L, "buu");
+        map2.put(10L, "flush-item");
+
+        Pipeline p = Pipeline.create();
+        StreamStage<Entry<Long, String>> stage1 = p.drawFrom(
+                Sources.<Long, String>mapJournal("source1", START_FROM_OLDEST));
+        stage1.addTimestamps(Entry::getKey, 0);
+
+        p.drawFrom(Sources.<Long, String>mapJournal("source", START_FROM_OLDEST))
+         .addTimestamps(Entry::getKey, 0)
+         .window(WindowDefinition.tumbling(2))
+         .aggregate2(stage1, toTwoBags())
+         .peek()
+         .drainTo(Sinks.list("sink"));
+
+        instance.newJob(p);
+        assertTrueEventually(() -> {
+            assertEquals(
+                    listToString(asList(
+                            new TimestampedItem<>(2, TwoBags.twoBags(
+                                    asList(entry(0L, "foo")),
+                                    asList(entry(0L, "faa"))
+                            )),
+                            new TimestampedItem<>(4, TwoBags.twoBags(
+                                    asList(entry(2L, "baz")),
+                                    asList(entry(2L, "buu"))
+                            )))),
+                    listToString(instance.getHazelcastInstance().getList("sink")));
+        }, 5);
+
+    }
+
+    @Test
+    public void test_aggregate3() {
+        IMap<Long, String> map = instance.getMap("source");
+        // key is timestamp
+        map.put(0L, "foo");
+        map.put(2L, "caz");
+        map.put(10L, "flush-item");
+
+        IMap<Long, String> map2 = instance.getMap("source1");
+        // key is timestamp
+        map2.put(0L, "faa");
+        map2.put(2L, "cuu");
+        map2.put(10L, "flush-item");
+
+
+        IMap<Long, String> map3 = instance.getMap("source2");
+        // key is timestamp
+        map3.put(0L, "fzz");
+        map3.put(2L, "ccc");
+        map3.put(10L, "flush-item");
+
+        Pipeline p = Pipeline.create();
+        StreamStage<Entry<Long, String>> stage1 = p.drawFrom(
+                Sources.<Long, String>mapJournal("source1", START_FROM_OLDEST));
+        stage1.addTimestamps(Entry::getKey, 0);
+        StreamStage<Entry<Long, String>> stage2 = p.drawFrom(
+                Sources.<Long, String>mapJournal("source2", START_FROM_OLDEST));
+        stage2.addTimestamps(Entry::getKey, 0);
+
+        p.drawFrom(Sources.<Long, String>mapJournal("source", START_FROM_OLDEST))
+         .addTimestamps(Entry::getKey, 0)
+         .window(WindowDefinition.tumbling(2))
+         .aggregate3(stage1, stage2, toThreeBags())
+         .peek()
+         .drainTo(Sinks.list("sink"));
+
+        instance.newJob(p);
+        assertTrueEventually(() -> {
+            assertEquals(
+                    listToString(asList(
+                            new TimestampedItem<>(2, ThreeBags.threeBags(
+                                    asList(entry(0L, "foo")),
+                                    asList(entry(0L, "faa")),
+                                    asList(entry(0L, "fzz"))
+                            )),
+                            new TimestampedItem<>(4, ThreeBags.threeBags(
+                                    asList(entry(2L, "caz")),
+                                    asList(entry(2L, "cuu")),
+                                    asList(entry(2L, "ccc"))
+                            )))),
+                    listToString(instance.getHazelcastInstance().getList("sink")));
+        }, 5);
+
     }
 }

